add announcement

Rup-Narayan-Rajbanshi opened the pull request:
## Changes
- Add Announcement model
- sync to firebase

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

Rup-Narayan-Rajbanshi committed:
chore(announcement): add announcement

tnagorra commented on apps/common/admin.py:
```diff
@@ -174,3 +176,16 @@ def get_readonly_fields(self, *args, **kwargs):
+
+@admin.register(Announcement)
+class AnnouncementAdmin(DjangoQLSearchMixin, FirebaseResourceAdmin, UserResourceAdmin, admin.ModelAdmin):
+    list_display = ("text", "is_active", "url")
+    readonly_fields = ("created_by", "modified_by")
```
Please check other models. Fields like created_by and modified_by are already defined in UserResourceAdmin

Rup-Narayan-Rajbanshi replied:
@tnagorra , https://github.com/mapswipe/mapswipe-backend/blob/07c00faf68d20e8c90b0bfd762ae92999107bb50/apps/common/admin.py#L28 Check here, created_by and modified by are commented. thus i added manually.

tnagorra replied:
In that case, we don't need to add created_by and modified_by in readonly_fields.

We can enable that from one place later.

tnagorra commented on apps/common/models.py:
```diff
@@ -305,3 +305,15 @@ def type_enum(self):
         return cls.objects.filter(marked_as_deleted=False)
+
+
+class Announcement(UserResource, FirebasePushResource):  # type: ignore[reportIncompatibleVariableOverride]
```
Not sure if this should be added on common/models
cc: @thenav56 ?

Rup-Narayan-Rajbanshi replied:
@tnagorra i didnot found any app relevant so i kept it in common/models.

tnagorra commented on apps/common/admin.py:
```diff
@@ -174,3 +176,16 @@ def get_readonly_fields(self, *args, **kwargs):
+        if obj.is_active:
+            Announcement.objects.exclude(id=obj.pk).update(is_active=False)
+        super().save_model(request, obj, form, change)
+        FirebaseAnnouncementPush(obj).trigger()
```
If we are de-activating an announcement, it should be deleted in firebase.

Rup-Narayan-Rajbanshi replied:
@tnagorra you mean, when we activate a announcement, all other announcement should be deleted from firebase.

kopitek8 replied:
The current behavior is that Firebase only stores one announcement (if that helps)

tnagorra replied:
@Rup-Narayan-Rajbanshi 
No. When there is no active announcement, we need to delete the announcement in firebase.

Rup-Narayan-Rajbanshi replied:
@kopitek8 Can we activate the announcement that is deactivated previously. In that case if Yes , we need to push to firebase again.

tnagorra replied:
Yes we can activate and deactivate the same announcement any number of time.

tnagorra commented on apps/common/admin.py:
```diff
@@ -174,3 +176,21 @@ def get_readonly_fields(self, *args, **kwargs):
+            previous_announcements = Announcement.objects.exclude(id=obj.id)
+            previous_announcements.update(is_active=False)
+            for announcement in previous_announcements:
+                FirebaseAnnouncementPush(announcement).handle_delete()
```
No need to delete as we are overwriting the existing announcement.

tnagorra committed:
fix(annoucement): use correct firebase uri for annoucement

- handle deletion using the FirebasePush.trigger function
- only enable explicit deletion on FirebasePush
- update example on announcement tests
- make migration
